### PR TITLE
Add ARM and s309 architectures to clock timing.

### PR DESCRIPTION
**Pull request by beberlei**
@andypost can you test these?

### CODE REVIEW

**Comment by andypost** [comment]
can't test on s390x right now but it breaks x86
```
>>> php7-tideways_xhprof: Unpacking /var/cache/distfiles/php7-tideways_xhprof-5.0_beta3.tar.gz...
>>> php7-tideways_xhprof: 73.patch
patching file timer.h
patching file timer.h
>>> php7-tideways_xhprof: 74.patch
patching file timer.h
Configuring for:
PHP Api Version:         20170718
Zend Module Api No:      20170718
Zend Extension Api No:   320170718
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for a sed that does not truncate output... /bin/sed
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking how to run the C preprocessor... gcc -E
checking for icc... no
checking for suncc... no
checking whether gcc and cc understand -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking target system type... x86_64-unknown-linux-gnu
checking for PHP prefix... /usr
checking for PHP includes... -I/usr/include/php7 -I/usr/include/php7/main -I/usr/include/php7/TSRM -I/usr/include/php7/Zend -I/usr/include/php7/ext -I/usr/include/php7/ext/date/lib
checking for PHP extension directory... /usr/lib/php7/modules
checking for PHP installed headers prefix... /usr/include/php7
checking if debug is enabled... no
checking if zts is enabled... no
checking for re2c... no
configure: WARNING: You will need re2c 0.13.4 or later if you want to regenerate PHP parsers.
checking for gawk... no
checking for nawk... no
checking for awk... awk
checking if awk is broken... no
checking whether to enable tideways_xhprof support... yes, shared
checking whether to enable developer build flags... no
checking for gettimeofday... yes
checking for clock_gettime... yes
checking for clock_gettime... yes
checking for ld used by gcc... /usr/i586-alpine-linux-musl/bin/ld
checking if the linker (/usr/i586-alpine-linux-musl/bin/ld) is GNU ld... yes
checking for /usr/i586-alpine-linux-musl/bin/ld option to reload object files... -r
checking for BSD-compatible nm... /usr/bin/nm -B
checking whether ln -s works... yes
checking how to recognize dependent libraries... pass_all
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking dlfcn.h usability... yes
checking dlfcn.h presence... yes
checking for dlfcn.h... yes
checking the maximum length of command line arguments... 98304
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for objdir... .libs
checking for ar... ar
checking for ranlib... ranlib
checking for strip... strip
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC
checking if gcc PIC flag -fPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking whether the gcc linker (/usr/i586-alpine-linux-musl/bin/ld -m elf_i386) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no

creating libtool
appending configuration tag "CXX" to libtool
configure: creating ./config.status
config.status: creating config.h
/bin/sh /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/libtool --mode=compile gcc -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1    -I. -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -DPHP_ATOM_INC -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/include -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/main -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -I/usr/include/php7 -I/usr/include/php7/main -I/usr/include/php7/TSRM -I/usr/include/php7/Zend -I/usr/include/php7/ext -I/usr/include/php7/ext/date/lib  -Os -fomit-frame-pointer -DHAVE_CONFIG_H  -Os -fomit-frame-pointer   -c /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tideways_xhprof.c -o tideways_xhprof.lo 
/bin/sh /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/libtool --mode=compile gcc -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1    -I. -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -DPHP_ATOM_INC -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/include -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/main -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -I/usr/include/php7 -I/usr/include/php7/main -I/usr/include/php7/TSRM -I/usr/include/php7/Zend -I/usr/include/php7/ext -I/usr/include/php7/ext/date/lib  -Os -fomit-frame-pointer -DHAVE_CONFIG_H  -Os -fomit-frame-pointer   -c /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.c -o tracing.lo 
mkdir .libs
 gcc -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1 -I. -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -DPHP_ATOM_INC -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/include -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/main -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -I/usr/include/php7 -I/usr/include/php7/main -I/usr/include/php7/TSRM -I/usr/include/php7/Zend -I/usr/include/php7/ext -I/usr/include/php7/ext/date/lib -Os -fomit-frame-pointer -DHAVE_CONFIG_H -Os -fomit-frame-pointer -c /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tideways_xhprof.c  -fPIC -DPIC -o .libs/tideways_xhprof.o
 gcc -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1 -I. -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -DPHP_ATOM_INC -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/include -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/main -I/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3 -I/usr/include/php7 -I/usr/include/php7/main -I/usr/include/php7/TSRM -I/usr/include/php7/Zend -I/usr/include/php7/ext -I/usr/include/php7/ext/date/lib -Os -fomit-frame-pointer -DHAVE_CONFIG_H -Os -fomit-frame-pointer -c /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.c  -fPIC -DPIC -o .libs/tracing.o
In file included from /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.h:1,
                 from /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tideways_xhprof.c:13:
/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h: In function 'time_milliseconds':
/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: error: 's' undeclared (first use in this function)
             if (clock_gettime(CLOCK_MONOTONIC, &s) == 0) {
                                                 ^
/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Makefile:194: tideways_xhprof.lo] Error 1
make: *** Waiting for unfinished jobs....
In file included from /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.h:1,
                 from /home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.c:11:
/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h: In function 'time_milliseconds':
/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: error: 's' undeclared (first use in this function)
             if (clock_gettime(CLOCK_MONOTONIC, &s) == 0) {
                                                 ^
/home/andypost/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Makefile:196: tracing.lo] Error 1
>>> ERROR: php7-tideways_xhprof: build failed
```

**Comment by andypost** [comment]
the same error on arm*

**Comment by andypost** [comment]
Still fails everywhere except x86-64 https://cloud.drone.io/alpinelinux/aports/2657

**Comment by andypost** [comment]
Today Alpine did switch to PHP 7.3 and build failed again for all arm/s390 

```
In file included from /home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.h:1,
                 from /home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.c:11:
/home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h: In function 'time_milliseconds':
/home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: error: 's' undeclared (first use in this function)
             if (clock_gettime(CLOCK_MONOTONIC, &s) == 0) {
                                                 ^
/home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: note: each undeclared identifier is reported only once for each function it appears in
In file included from /home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tracing.h:1,
                 from /home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/tideways_xhprof.c:13:
/home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h: In function 'time_milliseconds':
/home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: error: 's' undeclared (first use in this function)
             if (clock_gettime(CLOCK_MONOTONIC, &s) == 0) {
                                                 ^
/home/buildozer/aports/testing/php7-tideways_xhprof/src/php-xhprof-extension-5.0-beta3/timer.h:71:49: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Makefile:194: tideways_xhprof.lo] Error 1
```

**Comment by beberlei** [comment]
@andypost i have started over and pushed a cleaner version of the functions depending on arch. Can you check on ARM, s390 and i386? It works on x64 and windows.

**Comment by beberlei** [comment]
Also tested it on Linux + i386 successfully.

**Comment by andypost** [comment]
Thanks! I will test it tomorrow, btw I'm using locally https://github.com/alpinelinux/alpine-chroot-install to test other then x86-64 builds

File: timer.h [hunk 1]
```diff
@@ -9,36 +9,120 @@
...
+    uint32 a, d;
+    asm volatile("rdtsc" : "=a" (a), "=d" (d));
+    (val) = ((uint64)a) | (((uint64)d)<<32);
+    return val;
+#elif defined(__powerpc__) || defined(__ppc__)
+    uint64 val;
+    asm volatile ("mftb %0" : "=r" (val));
+    return val;
+#else
+    return 0;
```
Comment (andypost) [inline timer.h]: Does it mean that arm will use that result?

File: timer.h [hunk 2]
```diff
@@ -9,36 +9,120 @@
...
+        return s.tv_sec * 1000000 + s.tv_nsec / 1000;
+    }
+#endif
+
+    return 0;
+}
+
+static zend_always_inline uint64 time_milliseconds_gtod()
+{
+#if HAVE_GETTIMEOFDAY
```
Comment (andypost) [inline timer.h]: This will work for arm if hardware able?

File: timer.h (hunk 1 above)
Comment (beberlei) [reply timer.h]: Yes, ARM automatically uses GTOD by default and doesn't support TSC for now.

See https://github.com/tideways/php-xhprof-extension/pull/74/files#diff-9c31fd9ea16faaddb9f8d2d6a605e19bR28

File: timer.h (hunk 2 above)
Comment (beberlei) [reply timer.h]: I don't know, but it should work yes.

**Comment by andypost** [comment]
@beberlei Thanks for explanation, added whole diff from beta3 to MoreClocks branch to test it https://cloud.drone.io/alpinelinux/aports/9722/3/1
as you can see all ARMs fails on test `xhprof: dont disable no memory leak [tests/xhprof_006.phpt] `

That's because of weak test
```
tests$ cat xhprof_006.diff 
002+ Clock Source => gettimeofday
002- Clock Source => tsc
```

**Comment by beberlei** [comment]
@andypost fixed the test

**Comment by andypost** [comment]
@beberlei tested on arm and ppc64le - it works! the only remains are s390x (all tests fail)
[tests.zip](https://github.com/tideways/php-xhprof-extension/files/3514765/tests.zip)

**Comment by beberlei** [comment]
@andypost ok, s390x segfaults. Without a gdb backtrace I can't know why this happens. I would merge this now, but keep s390 out of the list of supported platforms.

**Comment by andypost** [comment]
Thanks a lot! it sounds like new beta could be released and address s390x in follow-up

PS: I just added dbg packages to Alpinelinux few days ago https://pkgs.alpinelinux.org/packages?name=php7-dbg&branch=edge

**Pull request merged**